frame/support/src/storage/mod.rs

```diff
@@ -811,16 +816,327 @@ mod private {
-/// we only encode the internal vec and we can append to this vec. We have a test that ensures
-/// that if the `Digest` format ever changes, we need to remove this here.
+/// We abuse the fact that SCALE does not put any marker into the encoding, i.e. we only encode the
+/// internal vec and we can append to this vec. We have a test that ensures that if the `Digest`
+/// format ever changes, we need to remove this here.
 impl<Hash: Encode> StorageAppend<DigestItem<Hash>> for Digest<Hash> {}
 
+/// Marker trait for types `T` that can be stored in storage as `Vec<T>`.
+pub trait Value: FullCodec + Clone + sp_std::fmt::Debug + Eq + PartialEq {}
+impl<T: FullCodec + Clone + sp_std::fmt::Debug + Eq + PartialEq> Value for T {}
+
+/// A bounded vector.
+///
+/// It implementations for efficient append and length decoding, as with a normal `Vec<_>`, once put
+/// into a [`StorageValue`].
+///
+/// As the name suggests, the length of the queue is always bounded. All internal operations ensure
+/// this bound is respected.
+#[derive(Encode, Decode, crate::DefaultNoBound, crate::CloneNoBound, crate::RuntimeDebugNoBound)]
+pub struct BoundedVec<T: Value, S: Get<u32>>(Vec<T>, sp_std::marker::PhantomData<S>);
+
+// NOTE: we could also implement this as:
+// impl<T: Value, S1: Get<u32>, S2: Get<u32>> PartialEq<BoundedVec<T, S2>> for BoundedVec<T, S1>
+// to allow comparison of bounded vectors with different bounds.
+impl<T: Value, S: Get<u32>> PartialEq for BoundedVec<T, S> {
+	fn eq(&self, rhs: &Self) -> bool {
+		self.0 == rhs.0
+	}
+}
+impl<T: Value, S: Get<u32>> Eq for BoundedVec<T, S> {}
+
+impl<T: Value, S: Get<u32>> BoundedVec<T, S> {
+	/// Create `Self` from `t` without any checks.
+	///
+	/// # WARNING
+	///
+	/// Only use when you are sure you know what you are doing.
+	fn unchecked_from(t: Vec<T>) -> Self {
+		Self(t, Default::default())
+	}
+
+	/// Create `Self` from `t` without any checks. Logs warnings if the bound is not being
+	/// respected. The additional scope can be used to indicate where a potential overflow is
+	/// happening.
+	///
+	/// # WARNING
+	///
+	/// Only use when you are sure you know what you are doing.
+	pub fn force_from(t: Vec<T>, scope: Option<&'static str>) -> Self {
+		if t.len() > Self::bound() {
+			log::warn!(
+				target: crate::LOG_TARGET,
+				"length of a bounded vector in scope {} is not respected.",
+				scope.unwrap_or("UNKNOWN"),
+			);
+		}
+
+		Self::unchecked_from(t)
+	}
+
+	/// Get the bound of the type in `usize`.
+	pub fn bound() -> usize {
+		S::get() as usize
+	}
+
+	/// Consume self, and return the inner `Vec`. Henceforth, the `Vec<_>` can be altered in an
+	/// arbitrary way. At some point, if the reverse conversion is required, `TryFrom<Vec<_>>` can
+	/// be used.
+	///
+	/// This is useful for cases if you need access to an internal API of the inner `Vec<_>` which
+	/// is not provided by the wrapper `BoundedVec`.
+	pub fn inner(self) -> Vec<T> {
```

Comment:
```suggestion
	pub fn into_inner(self) -> Vec<T> {
```

However, did you thought of using `Deref` + overwriting the methods that should not be accessible without bound checking? This should be only `push` and `extend`?

Reply:
Not really, my initial thought process was to make the API restrictive; rather than providing everything except those that might extend the length, I decided on the opposite. 

But I see a lot of benefit in what you propose, and it is simpler. Will look into it.

Reply:
Ahh wait! I'm dumb and smart at once xD By just providing `Deref`, you will only give access to non-mutable functions. Meaning, nothing that would add something to the vec would be allowed :D

Reply:
💡🧠

Reply:
Applied and much nicer; I provide some functions that are safe with `&mut` sich as `remove`, `retain` etc, and just a `Deref` to get everything else.